CMakeLists.txt

```diff
@@ -203,18 +203,20 @@ else()
 endif()
 
 include(CheckCCompilerFlag)
+include(CheckCXXSourceCompiles)
 if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)64")
-  CHECK_C_COMPILER_FLAG("-mcpu=power9" HAS_POWER9)
-  if(HAS_POWER9)
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -mcpu=power9 -mtune=power9")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mcpu=power9 -mtune=power9")
-  else()
-    CHECK_C_COMPILER_FLAG("-mcpu=power8" HAS_POWER8)
-    if(HAS_POWER8)
-      set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -mcpu=power8 -mtune=power8")
-      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mcpu=power8 -mtune=power8")
-    endif(HAS_POWER8)
-  endif(HAS_POWER9)
+  CHECK_CXX_SOURCE_COMPILES("
+  #ifndef _ARCH_PWR8
+  #error Not POWER8+
+  #endif
+  int main() { return 0; }"
+  HAS_POWER8)
+  if(NOT CMAKE_C_FLAGS MATCHES "m(cpu|tune|arch)")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -mcpu=native")
```

Comment:
From the [GCC docs](https://gcc.gnu.org/onlinedocs/gcc/RS_002f6000-and-PowerPC-Options.html):

> Specifying ‘native’ as cpu type detects and selects the architecture option that corresponds to the host processor of the system performing the compilation. -mcpu=native has no effect if GCC does not recognize the processor.

> The other options specify a specific processor. Code generated under those options runs best on that processor, and may not run at all on others.

So if I understand correctly:
Previously... if building on a POWER9 system then the binaries were optimised for POWER9 ISA. If building on a POWER8 the binaries should have been optimised for POWER8, but the HAS_POWER9 flag was being set to true on a POWER8 and the resultant compiled binaries would not run on said POWER8.

---

I think we need the option to be able to build CPU optimised and non-optimised binaries. For portability practices we already have the `PORTABLE` flag which is used in Makefile and also in CMakeLists.txt.

1. For optimised builds, I think we should combine what we had previously, with your extra CHECK_CXX_SOURCE_COMPILES check on POWER8 to ensure that POWER8 compiled binaries will run on Power8. 

2. For portable binaries, I think that on ppc64/le we should check the presence of the PORTABLE flag and should likely always set a minimum system ISA. Whether that should be POWER8 or something earlier like POWER6 I can't say at the moment... I am not certain what is representative of the installed base of PPC systems out there.

When performing release builds we already set the PORTABLE flag, so this should provide what you need.

In addition, whatever changes we make to CMakeLists.txt should also be similarly reflected in our Makefile setup too.

Reply:
yes you've understood it correctly. Good idea making more use of `PORTABLE`.

1. native(?) optimised

What was there previously was forcing the compile cpu to POWER9, which isn't going to run on POWER8. The CHECK_CXX_SOURCE_COMPILES ensures a minimum of POWER8. This is why I set mcpu to native. So on P8 we'd get P8 code and on P9 we'd get P9. Do you have a good runtime check for cmake for P8? `grep /proc/cpuinfo`  but that is bit ugly. I'm not sure what this gains over mcpu=native. Maybe I've misunderstood something. Apologies if I have.

2. Portable:

I can check on a POWER7. POWER6 might be harder to find. So sure, can make PORTABLE default to the minimum workable architecture.

2.1 portable, but still optimized.

For functions like crc32 that have a significant performance difference we could use https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#Common-Function-Attributes - target_clones maybe? Any other functions you've noticed significant p8 vs p9 gains in?

Reply:
> Do you have a good runtime check for cmake for P8? grep /proc/cpuinfo but that is bit ugly. I'm not sure what this gains over mcpu=native. Maybe I've misunderstood something. Apologies if I have.

I think I am also getting a bit confused. My concern I think was that we were previously setting both `-mcpu` and -`mtune` however we are now only setting `-mcpu`... https://gcc.gnu.org/onlinedocs/gcc/RS_002f6000-and-PowerPC-Options.html

> So sure, can make PORTABLE default to the minimum workable architecture.

Excellent!

> Any other functions you've noticed significant p8 vs p9 gains in?

I don't have enough experience with using Rocks on ppc64le systems to really comment on this. Most of my work was just making sure that it would compile on ppc64le.